packages/net/src/provider/non_blocking.rs

```diff
@@ -0,0 +1,117 @@
+use crate::{NetProvider, RequestHandler, USER_AGENT};
+use data_url::DataUrl;
+use futures_util::{stream::FuturesUnordered, StreamExt};
+use reqwest::Client;
+use std::{sync::Arc, time::Duration};
+use thiserror::Error;
+use tokio::{
+    runtime::{Handle, Runtime},
+    sync::Mutex,
+    task::JoinHandle,
+};
+use url::Url;
+use winit::{event_loop::EventLoopProxy, window::WindowId};
+
+type TaskHandle<T> = JoinHandle<Result<T, AsyncProviderError>>;
+
+pub struct AsyncProvider<T> {
+    rt: Handle,
+    client: Client,
+    futures: Mutex<FuturesUnordered<TaskHandle<T>>>,
+}
+impl<T: Send + Sync + 'static> AsyncProvider<T> {
+    pub fn new(rt: &Runtime) -> Self {
+        Self {
+            rt: rt.handle().clone(),
+            client: Client::new(),
+            futures: Mutex::new(FuturesUnordered::new()),
+        }
+    }
+    pub fn resolve<P: From<(WindowId, T)> + Send>(
```

Comment:
If you have `fetch` accept the `WindowId` and the callback for the result directly, then I think this resolve function can be removed.

Reply:
The problem is that basically all fetching is done before the window (and the WindowId) exist.

Reply:
I think this could be solved by giving the Document itself an ID The ID (perhaps just a simple sequential one?) could be generated either by the high-level glue code (or by `Document::new()`) and then the high-level code could keep track of which windows/documents that ID corresponds to.

This seems like a good idea anyway as in future we'll likely want to support things like replacing the Document associated with a Window (e.g. when clicking on a link), and having sub-Documents within a Document (e.g. for iframes, or using DioxusDocument wrapping an HtmlDocument to implement a browser ui).